Python/Product/Analysis/LanguageServer/ServerBase.cs

```diff
@@ -0,0 +1,232 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    public abstract class ServerBase {
+        private RequestLock _lock;
+
+        private sealed class RequestLock : IDisposable {
+            public readonly ServerBase Owner;
+            public readonly CancellationTokenSource CancellationTokenSource;
+
+            public CancellationToken Token => CancellationTokenSource.Token;
+            public void Cancel() => CancellationTokenSource.Cancel();
```

Comment:
If `Cancel` is called after `Dispose`, there will be `ObjectDisposedException`

Reply:
Yes. You shouldn't call `CancelRequest` outside of an `AllowRequestCancellation` state or your object has been disposed.

Reply:
Ok